storage/db.go

```diff
@@ -0,0 +1,500 @@
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package storage
+
+import (
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/boltdb/bolt"
+	"github.com/katzenpost/client/constants"
+	"github.com/katzenpost/client/crypto/block"
+	sphinxconstants "github.com/katzenpost/core/sphinx/constants"
+)
+
+const (
+	// BlockIDLength is the length of our storage block IDs
+	// which are used to uniquely identify storage blocks
+	// in the boltdb ingress buckets
+	BlockIDLength = 8
+
+	// EgressBucketName is the name of the boltdb bucket
+	// used for storing messages received from our SMTP listener.
+	// We intentionally have a single boltdb bucket that handles
+	// all the outgoing messages for the client.
+	EgressBucketName = "outgoing"
+)
+
+// ingressBucketNameFromAccount is a helper function that
+// returns the bucket name of the bucket that persists
+// encrypted message blocks given the name of an account.
+// (in this case the account is an e-mail address)
+func ingressBucketNameFromAccount(accountName string) []byte {
+	return []byte(fmt.Sprintf("%s_incoming", accountName))
+}
+
+// pop3BucketNameFromAccount is a helper function that
+// returns the bucket name of the bucket that persists
+// plaintext message constructed from one or more
+// encrypted blocks from the account's "_incoming" bucket.
+func pop3BucketNameFromAccount(accountName string) []byte {
+	return []byte(fmt.Sprintf("%s_pop3", accountName))
+}
+
+// StorageBlock contains an encrypted message fragment
+// and other fields needed to send it to the destination
+type StorageBlock struct {
+
+	// BlockID is used to uniquely identify storage blocks
+	BlockID [BlockIDLength]byte
+
+	// Sender is the sender identity (aka e-mail address)
+	Sender string
+
+	// SenderProvider is the Provider for a given sender.
+	// (the part of the email address after the @-sign)
+	SenderProvider string
+
+	// Recipient is the recipient identity/e-mail address
+	Recipient string
+
+	// RecipientProvider is the Provider name of the recipient
+	// (the part of the email address after the @-sign)
+	RecipientProvider string
+
+	// RecipientID is the user ID for a given recipient
+	// which is padded to fixed length
+	RecipientID [sphinxconstants.RecipientIDLength]byte
+
+	// SendAttempts is the number of attempts to retransmit
+	// a given message block
+	SendAttempts uint8
+
+	// SURBKeys are the keys used to decrypt a message
+	// composed using a SURB. See github.com/katzenpost/core/sphinx
+	SURBKeys []byte
+
+	// SURBID is used to uniquely identify a message and decryption keys
+	// for a message composed using a SURB.
+	SURBID [sphinxconstants.SURBIDLength]byte
+
+	// Block is a message fragment
+	Block block.Block
+}
+
+// JsonStorageBlock is a json serializable representation of StorageBlock
+type JsonStorageBlock struct {
```

Comment:
Does this type need to be exported, or can it be lower-cased / package private?

Reply:
unclear. i will investigate.